File: qc-core/src/main/java/com/eager/questioncloud/core/domain/point/implement/ChargePointPaymentFailHandler.java
package com.eager.questioncloud.core.domain.point.implement;

import com.eager.questioncloud.core.domain.pg.PGAPI;
import com.eager.questioncloud.core.domain.point.model.ChargePointPayment;
import com.eager.questioncloud.core.domain.point.repository.ChargePointPaymentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class ChargePointPaymentFailHandler {
    private final ChargePointPaymentRepository chargePointPaymentRepository;
    private final PGAPI pgAPI;

    public void failHandler(String paymentId) {
        ChargePointPayment chargePointPayment = chargePointPaymentRepository.findByPaymentId(paymentId);
        chargePointPayment.fail();
        chargePointPaymentRepository.save(chargePointPayment);

        pgAPI.cancel(chargePointPayment.getPaymentId());
    }
}

File: qc-core/qc-point/qc-point-api/src/main/kotlin/com/eager/questioncloud/point/implement/ChargePointPaymentFailHandler.kt
package com.eager.questioncloud.point.implement

import com.eager.questioncloud.point.domain.ChargePointPaymentIdempotentInfo
import com.eager.questioncloud.point.enums.ChargePointPaymentStatus
import com.eager.questioncloud.point.repository.ChargePointPaymentIdempotentInfoRepository
import com.eager.questioncloud.point.repository.ChargePointPaymentRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class ChargePointPaymentFailHandler(
    private val chargePointPaymentRepository: ChargePointPaymentRepository,
    private val chargePointPaymentIdempotentInfoRepository: ChargePointPaymentIdempotentInfoRepository
) {
    @Transactional
    fun fail(orderId: String, paymentId: String) {
        val chargePointPayment = chargePointPaymentRepository.findByOrderId(orderId)
        val idempotentInfo = ChargePointPaymentIdempotentInfo(
            orderId = orderId,
            paymentId = paymentId,
            chargePointPaymentStatus = ChargePointPaymentStatus.FAILED
        )
        
        if (chargePointPaymentIdempotentInfoRepository.insert(idempotentInfo)) {
            chargePointPayment.fail()
            chargePointPaymentRepository.update(chargePointPayment)
        }
    }
}